### PR TITLE
refactor(turborepo): Two binaries

### PR DESCRIPTION
**Pull request by NicholasLYang**
Because of golang/go#13492, we cannot use Go as a shared library that we link to Rust. Instead we are now building the Go code as a separate binary that is executed by the Rust one.

### CODE REVIEW

**Review by gsoltis — APPROVED** [review]

File: cli/Makefile [hunk 1]
```diff
@@ -19,14 +21,11 @@ GO_FILES = $(shell find . -name "*.go")
...
+turbo: go-turbo$(EXT)
 	cargo build --manifest-path ../crates/turborepo/Cargo.toml
 
-libturbo.a: $(GENERATED_FILES) $(SRC_FILES) go.mod
-	go build -buildmode=c-archive -o libturbo.a  ./cmd/turbo/...
-
-turbo.lib: $(GENERATED_FILES) $(SRC_FILES) go.mod
-	go build -buildmode=c-archive -o turbo.lib  ./cmd/turbo/...
+go-turbo$(EXT): $(GENERATED_FILES) $(SRC_FILES) go.mod
+	go build $(GO_FLAGS) ./cmd/turbo
```
Comment (gsoltis) [inline cli/Makefile:28]:
```suggestion
	CGO_ENABLED=1 go build $(GO_FLAGS) -o go-turbo$(EXT) ./cmd/turbo
```

**Review by gsoltis — COMMENTED** [review]
`Makefile` is still missing some changes. We need to include all of the `build-go-turbo-*` targets

File: crates/turborepo/src/main.rs [hunk 2]
```diff
@@ -1,32 +1,29 @@
...
 use turborepo_lib::{Args, Payload};
 
-use crate::ffi::{nativeRunWithArgs, GoString};
+fn run_go_binary(args: Args) -> Result<i32> {
+    let mut go_binary_path = current_exe()?;
+    go_binary_path.pop();
+    #[cfg(windows)]
+    {
+        go_binary_path.push("go-turbo.exe");
+    }
```
Comment (NicholasLYang) [inline crates/turborepo/src/main.rs:12]:
```suggestion
        go_binary_path.push("go-turbo.exe");
```

File: crates/turborepo/src/main.rs [hunk 3]
```diff
@@ -1,32 +1,27 @@
...
 
 use anyhow::Result;
 use turborepo_lib::{Args, Payload};
 
-use crate::ffi::{nativeRunWithArgs, GoString};
+fn run_go_binary(args: Args) -> Result<i32> {
+    let mut go_binary_path = current_exe()?;
+    go_binary_path.pop();
+    #[cfg(windows)]
+        go_binary_path.push("go-turbo.exe");
```
Comment (gsoltis) [inline crates/turborepo/src/main.rs:10]:
```suggestion
    go_binary_path.push("go-turbo.exe");
```

File: crates/turborepo/build.rs [hunk 4]
```diff
@@ -1,77 +1,13 @@
...
-    let out_path = Path::new(&out_dir).join("bindings.rs");
-
-    bindings
-        .write_to_file(out_path)
-        .expect("Couldn't write bindings!");
-
-    if target.os == build_target::Os::MacOs {
-        println!("cargo:rustc-link-lib=framework=cocoa");
-        println!("cargo:rustc-link-lib=framework=security");
+    let is_ci_release = matches!(env::var("PROFILE"), Ok(profile) if profile == "release");
```
Comment (ForsakenHarmony) [inline crates/turborepo/build.rs:4]: This means running `cargo run --release` will break, is that intentional?

Comment (gsoltis) [reply crates/turborepo/build.rs:4]: Running `cargo run --release` won't break, it just doesn't build the Go binary. Our release process builds the Go binary separately, setting a bunch of values for each supported platform.

Comment (ForsakenHarmony) [reply crates/turborepo/build.rs:4]: Well won't break, but it also won't be able to run

File: crates/turborepo/build.rs [hunk 5]
```diff
@@ -90,21 +26,38 @@ fn build_debug_libturbo() -> PathBuf {
         for ext in ["pdb", "exe", "d", "lib"].iter() {
```
Comment (ForsakenHarmony) [inline crates/turborepo/build.rs:26]: This probably isn't necessary anymore

File: crates/turborepo/build.rs (hunk 4 above)
Comment (gsoltis) [reply crates/turborepo/build.rs:4]: Correct, if you want to run a release build of the rust binary, you are responsible for providing the Go binary outside of the rust build process. 

For development work, the debug build still calls over to the `Makefile` to build the Go binary for you.

File: crates/turborepo/build.rs [hunk 6]
```diff
@@ -90,21 +26,38 @@ fn build_debug_libturbo() -> PathBuf {
...
+        .join(go_binary_name);
+
+    let new_go_binary_path = env::var_os("CARGO_WORKSPACE_DIR")
+        .map(PathBuf::from)
+        .unwrap()
+        .join("target")
+        .join("debug")
+        .join(go_binary_name);
+
+    fs::rename(go_binary_path, new_go_binary_path).unwrap();
```
Comment (ForsakenHarmony) [inline crates/turborepo/build.rs:60]: Wouldn't copying be better?

Also, might want to add a `cargo:rerun-if-changed` to pick up changes

File: crates/turborepo/build.rs (hunk 4 above)
Comment (NicholasLYang) [reply crates/turborepo/build.rs:4]: We could keep the `RELEASE_TURBO_CLI` variable to only not build the Go binary if we're running in CI

File: crates/turborepo/build.rs (hunk 5 above)
Comment (NicholasLYang) [reply crates/turborepo/build.rs:26]: Not a windows expert here. @chris-olszewski confirmed?

Comment (chris-olszewski) [reply crates/turborepo/build.rs:30]: I believe that's correct.

**Review by tknickman — APPROVED** [review]

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**